EventListener/BreadcrumbListener.php

```diff
@@ -0,0 +1,402 @@
+    {
+        $controller = $event->getController();
+        if (!is_array($controller)) {
+            return;
+        }
+
+        if (HttpKernelInterface::MASTER_REQUEST == $event->getRequestType()) {
+            $this->breadcrumbTrail->reset();
+        }
+
+        $this->processAnnotations($event, $controller);
+    }
+
+    private function processAnnotations(KernelEvent $event, array $controller): void
+    {
+        $class = new ReflectionClass($controller[0]);
+        if ($class->isAbstract()) {
+            throw new InvalidArgumentException(sprintf('Annotations from class "%s" cannot be read as it is abstract.', $class));
+        }
+
+        $method = $class->getMethod($controller[1]);
+        if (HttpKernelInterface::MASTER_REQUEST == $event->getRequestType()) {
+            $this->addBreadcrumbsFromClass($event, $this->reader->getClassAnnotations($class));
+            $this->addBreadcrumbsFromMethod($event, $this->reader->getMethodAnnotations($method));
+        }
+    }
+
+    private function processParentAnnotations(
+        KernelEvent $event,
+        array $controller,
+        ?string $routeName = null,
+        ?array $parameters = null
+    ): void {
+        $class = new ReflectionClass($controller[0]);
+        $method = $class->getMethod($controller[1]);
+
+        if (HttpKernelInterface::MASTER_REQUEST == $event->getRequestType()) {
+            $this->addBreadcrumbsFromMethod(
+                $event,
+                $this->reader->getMethodAnnotations($method),
+                $routeName,
+                $parameters
+            );
+        }
+    }
+
+    private function addBreadcrumbsFromMethod(
+        KernelEvent $event,
+        array $annotations,
+        ?string $routeName = null,
+        ?array $parameters = null
+    ): void {
+        $this->addBreadcrumbsFromAnnotations($event, $annotations, $routeName, $parameters);
+    }
+
+    private function addBreadcrumbsFromClass(KernelEvent $event, array $annotations): void
+    {
+        $this->addBreadcrumbsFromAnnotations($event, $annotations);
+    }
+
+    private function addBreadcrumbsFromAnnotations(
+        KernelEvent $event,
+        array $annotations,
+        ?string $routeName = null,
+        ?array $parameters = null
+    ) {
+        foreach ($annotations as $annotation) {
+            if (!$annotation instanceof Breadcrumb) {
+                continue;
+            }
+
+            if ($annotation->getParentRouteName()) {
+                $this->addBreadcrumbsForParent($annotation, $event);
+            }
+
+            $this->breadcrumbTrail->add(
+                $this->generateBreadcrumb(
+                    $event->getRequest(),
+                    $annotation,
+                    $routeName,
+                    $parameters
+                )
+            );
+        }
+    }
+
+    private function generateBreadcrumb(
+        Request $request,
+        Breadcrumb $breadcrumb,
```

Comment:
Ergens vind ik het wat raar dat de annotation class `Breadcrumb` is.  Hierboven doen we
```php
$this->generateBreadcrumb(
      $event->getRequest(),
      $annotation, <--
      $routeName,
      $parameters
)
```
en die `$annotation` bevat dan een Breadcrumb class? In mijn hoofd klinkt BreadcrumbAnnotation daar beter. Ik zou dan ook eerder `BreadcrumbValueObject` eigenlijk een `Breadcrumb` maken. Zo heeft `generateBreadcrumb` ook `Breadcrumb` als return type, wat logisch is.

Reply:
fixed 193ad69842ccecda6c12c40c96ae9618bda5b720